File: cpp/src/bws/packer/PackerForm.cpp
#include <bws/packer/PackerForm.hpp>

#include <bws/packer/InstanceFormArray.hpp>
#include <bws/packer/WrapperArray.hpp>
#include <bws/packer/Packer.hpp>
#include <samchon/library/GAParameters.hpp>

using namespace std;
using namespace bws::packer;
using namespace samchon::library;
using namespace samchon::protocol;

/* -----------------------------------------------------------
	CONSTRUCTORS
----------------------------------------------------------- */
PackerForm::PackerForm()
	: super()
{
	instanceFormArray.reset(new InstanceFormArray());
	wrapperArray.reset(new WrapperArray());
	gaParameters.reset(new GAParameters());
}

void PackerForm::construct(shared_ptr<XML> xml)
{
	instanceFormArray->construct(xml->get(instanceFormArray->TAG())->at(0));
	wrapperArray->construct(xml->get(wrapperArray->TAG())->at(0));
	
	if (xml->has(gaParameters->TAG()))
		gaParameters->construct(xml->get(gaParameters->TAG())->at(0));
	else
		gaParameters.reset(new GAParameters());
}

/* -----------------------------------------------------------
	ACCESSORS
----------------------------------------------------------- */
auto PackerForm::optimize() const -> shared_ptr<WrapperArray>
{
	auto &packer = this->toPacker();

	return packer->optimize(*gaParameters);
}

auto PackerForm::getGAParameters() const -> shared_ptr<GAParameters>
{
	return gaParameters;
}

/* -----------------------------------------------------------
	EXPORTERS
----------------------------------------------------------- */
auto PackerForm::TAG() const -> string
{
	return "packerForm";
}

auto PackerForm::toXML() const -> shared_ptr<XML>
{
	shared_ptr<XML> &xml = super::toXML();
	xml->push_back(instanceFormArray->toXML());
	xml->push_back(wrapperArray->toXML());
	xml->push_back(gaParameters->toXML());

	return xml;
}
auto PackerForm::toPacker() const -> shared_ptr<Packer>
{
	return make_shared<Packer>(wrapperArray, instanceFormArray->toInstanceArray());
}

File: ts/src/bws/packer/PackerForm.ts
namespace bws.packer
{
	/**
	 * Bridge of {@link Packer} for {@link InstanceForm repeated instances}.
	 *
	 * @author Jeongho Nam <http://samchon.org>
	 */
	export class PackerForm
		extends samchon.protocol.Entity
	{
		private instanceFormArray: InstanceFormArray;
		private wrapperArray: WrapperArray;

		/* -----------------------------------------------------------
			CONSTRUCTORS
		----------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();

			this.instanceFormArray = new InstanceFormArray();
			this.wrapperArray = new WrapperArray();
		}
		public construct(xml: samchon.library.XML): void
		{
			this.instanceFormArray.construct(xml.get(this.instanceFormArray.TAG()).at(0));
			this.wrapperArray.construct(xml.get(this.wrapperArray.TAG()).at(0));
		}

		/* -----------------------------------------------------------
			ACCESSORS
		----------------------------------------------------------- */
		public optimize(): WrapperArray
		{
			let packer: Packer = this.toPacker();

			return packer.optimize();
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		public TAG(): string
		{
			return "packerForm";
		}
		public toXML(): samchon.library.XML
		{
			let xml: samchon.library.XML = super.toXML();
			xml.push(this.instanceFormArray.toXML());
			xml.push(this.wrapperArray.toXML());

			return xml;
		}

		public toPacker(): Packer
		{
			let packer: Packer = new Packer(this.wrapperArray, this.instanceFormArray.toInstanceArray());
			return packer;
		}
	}

	/**
	 * An array of {@link InstanceForm} objects.
	 *
	 * @author Jeongho Nam <http://samchon.org>
	 */
	class InstanceFormArray
		extends samchon.protocol.EntityArray<InstanceForm>
	{
		/* -----------------------------------------------------------
			CONSTRUCTORS
		----------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();
		}
		protected createChild(xml: samchon.library.XML): InstanceForm
		{
			return new InstanceForm();
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		public TAG(): string
		{
			return "instanceFormArray";
		}
		public CHILD_TAG(): string
		{
			return "instanceForm";
		}

		/**
		 * Convert {@link InstanceForm} objects to {@link InstanceArray}.
		 *
		 * @return An array of instance containing repeated instances in {@link InstanceForm} objects.
		 */
		public toInstanceArray(): InstanceArray
		{
			let instanceArray: InstanceArray = new InstanceArray();
			for (let i: number = 0; i < this.size(); i++)
			{
				let myInstances: InstanceArray = this.at(i).toInstanceArray();
				//for (let j: number = 0; j < myInstances.size(); j++)
				//	instanceArray.pushBack(myInstances.at(j));


				instanceArray.insert(instanceArray.end(), myInstances.begin(), myInstances.end());
			}

			return instanceArray;
		}
	}


	/**
	 * <p> A repeated Instance. </p>
	 *
	 * <p> InstanceForm is an utility class for repeated {@link Instance}. It is designed for shrinking 
	 * volume of network message I/O by storing {@link count repeated count}. </p>
	 *
	 * @author Jeongho Nam <http://samchon.org>
	 */
	class InstanceForm
		extends samchon.protocol.Entity
	{
		/**
		 * A duplicated Instance.
		 */
		private instance: Instance;

		/**
		 * Repeated count of the {@link instance}.
		 */
		private count: number;

		/* -----------------------------------------------------------
			CONSTRUCTORS
		----------------------------------------------------------- */
		/**
		 * Default Constructor.
		 */
		public constructor()
		{
			super();

			this.instance = null;
			this.count = 1;
		}

		public construct(xml: samchon.library.XML): void
		{
			super.construct(xml);

			if (xml.hasProperty("type"))
			{
				this.instance = this.createInstance(xml);
				this.instance.construct(xml);
			}
			else if (xml.has("instance"))
			{
				let instanceXML: samchon.library.XML = xml.get("instance").at(0);

				this.instance = this.createInstance(instanceXML);
				this.instance.construct(instanceXML);
			}
		}

		private createInstance(xml: samchon.library.XML): Instance
		{
			if (xml.getProperty("type") == "product")
				return new Product();
			else
				return new Wrapper();
		}

		/* -----------------------------------------------------------
			EXPORTERS
		----------------------------------------------------------- */
		public TAG(): string
		{
			return "instanceForm";
		}
		public toXML(): samchon.library.XML
		{
			let xml = super.toXML();

			if (this.instance != null)
				xml.push(this.instance.toXML());

			return xml;
		}

		/**
		 * <p> Repeated {@link instance} to {@link InstanceArray}.
		 *
		 * @details 
		 * <p> Contains the {@link instance repeated instance} to an {@link InstanceArray} to make 
		 * {@link instance} to participate in the packing process. The returned {@link InstanceArray} will be 
		 * registered on {@link Packer.instanceArray}.
		 *
		 * @return An array of instance containing repeated {@link instance}.
		 */
		public toInstanceArray(): InstanceArray
		{
			let instanceArray: InstanceArray = new InstanceArray();
			instanceArray.assign(this.count, this.instance);
			
			return instanceArray;
		}
	}
}